### PR TITLE
Change dataflow and nificluster controllers to avoid changing status on every reconciliation loop 

### LINKED ISSUE #119: NifiCluster reconciliation loop is non-idempotent in ready state
### Basic description
Once a `NifiCluster` has been fully reconciled--all pods created and ready, users/groups have certificates, data flows have been added, etc.--the `NifiCluster` resource continues to be modified by every single iteration of the reconciliation loop. This results in another iteration of the loop being queued, ad nauseum.

This is typically the result of a non-idempotency bug ([example](https://github.com/kubernetes-sigs/controller-runtime/issues/1613)). Some part of the work being done by the [`Reconcile`](https://github.com/konpyutaika/nifikop/blob/master/controllers/nificluster_controller.go#L81) function is causing the resource to be constantly updated.

### Evidence
This can easily be observed in two ways. First, run the following command on a `NifiCluster` once all child resources have been fully reconciled:
```bash
kubectl get nificluster <cluster-name> -o yaml --watch
```

You will see the `status.state` field flickering between `ClusterReconciling` and `ClusterRunning` like so:

nificluster-status-flickering

Alternatively, you can observe the `resourceVersion` of the object constantly--and rapidly--being incremented. Here's one command that demonstrates this using [`yq`](https://github.com/mikefarah/yq) to extract the field from the yaml output we are `--watch`ing:
```bash
kubectl get nificluster <cluster-name> -o yaml --watch | yq '.metadata.resourceVersion'
```
nificluster-resourceversion-incrementing

### Impact
This bug causes pain in several areas. First, is in how GitOps projects like ArgoCD interact with the `NifiCluster` custom resource and some of its children. Because the `NifiCluster` resource is constantly changing, ArgoCD believes that the resource is constantly progressing. Take a look:

nificluster-argocd

Another common result of this issue is error messages from the k8s api when attempting to modify the `NifiCluster` resource. The error explains that a change to the resource could not be fulfilled due to the fact that the resource was modified in the middle of crafting the request, and is thus no longer the latest version of the resource. These messages can be seen in `nifikop` logs: 

Screenshot_64

### Additional context
This bug occurs in the latest version of nifikop and for all versions of NiFi. 

There was a slack discussion regarding this bug [https://konpytika.slack.com/archives/C035X6KP684/p1649863690965119](https://konpytika.slack.com/archives/C035X6KP684/p1649863690965119).

### PR DESCRIPTION
**Pull request by michael81877**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | no
| API breaks?     | no
| Deprecations?   | no
| Related tickets | fixes #119 
| License         | Apache 2.0

### What's in this PR?

Tweaks `NifiCluster` and `NifiDataflow` controllers to avoid changing their respective statuses on every reconciliation loop. Instead, the controllers will only change the status to ready/done and log any actions taken thereafter and they will respect their reconciliation interval settings (15s by default). Previously to this, the controllers would reconcile many times _per second_, which drastically increases the load on the control plane and on NiFi itself.

Additionally, i changed the logger timestamp encoder to user ISO8601 instead of epoch time so it is human readable.

### Why?

Issue #119 does an excellent job of explaining why this is an issue.

### Checklist

- [x] Implementation tested
- [x] Error handling code meets the [guideline](docs/error-handling-guide.md)
- [x] Logging code meets the guideline
- [x] User guide and development docs updated (if needed)
- [x] Append changelog with changes

### CODE REVIEW

File: controllers/nificluster_controller.go [hunk 1]
```diff
@@ -181,11 +175,13 @@ func (r *NifiClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
-	if err := k8sutil.UpdateCRStatus(r.Client, instance, v1alpha1.NifiClusterRunning, r.Log); err != nil {
-		return RequeueWithError(r.Log, err.Error(), err)
+	if !instance.IsReady() {
+		if err := k8sutil.UpdateCRStatus(r.Client, instance, v1alpha1.NifiClusterRunning, r.Log); err != nil {
+			return RequeueWithError(r.Log, err.Error(), err)
+		}
 	}
 
-	return Reconciled()
+	return RequeueAfter(intervalRunning)
```
Comment (michael81877) [inline controllers/nificluster_controller.go:184]: @cannonpalms asked:

> Can you speak to the strategy for watching child resources (namely pods) now that the NifiCluster controller is not flapping?

Prior to this change, the `NifiCluster` controller perpetually scheduled reconciliation requests by updating its own status (`Reconciling` vs `Running`). The controller still watches all of the resources it was watching before. The difference is that now the controller schedules reconciliation requests respecting its reconciliation interval of `15s` -- that's the change on this line of code. Reconciliation requests will also be triggered if the `NifiCluster` CRD is modified by a user.

The same is true for the `NifiDataflow` controller.

**Review by erdrix — CHANGES_REQUESTED** [review]
It would be nice to replace the deleted state change with an event notification to keep the information, but in a better place ?

**Comment by michael81877** [comment]
> It would be nice to replace the deleted state change with an event notification to keep the information, but in a better place ?

Definitely! Let me add an event there instead

**Comment by michael81877** [comment]
@erdrix I've made sure an event notification and log message are generated when the controllers begin & end reconciliation 👍

**Comment by r65535** [comment]
LGTM!
@erdrix - was there a reason for adding `RequeueAfter(interval / 3)` to the dataflow controller [here](https://github.com/konpyutaika/nifikop/blame/49a9b5e1ac8d596bb388a43bdedeb0135923c7d3/controllers/nifidataflow_controller.go#L471)?

**Comment by erdrix** [comment]
> LGTM! @erdrix - was there a reason for adding `RequeueAfter(interval / 3)` to the dataflow controller [here](https://github.com/konpyutaika/nifikop/blame/49a9b5e1ac8d596bb388a43bdedeb0135923c7d3/controllers/nifidataflow_controller.go#L471)?

If so, I can't remember why...

**Review by erdrix — APPROVED** [review]

**Comment by erdrix** [comment]
@cannonpalms, can you confirm that this solves some of your problems?

**Comment by michael81877** [comment]
> > LGTM! @erdrix - was there a reason for adding `RequeueAfter(interval / 3)` to the dataflow controller [here](https://github.com/konpyutaika/nifikop/blame/49a9b5e1ac8d596bb388a43bdedeb0135923c7d3/controllers/nifidataflow_controller.go#L471)?
> 
> If so, I can't remember why...

Since the dataflow controller will now respect its interval, should we either revert the division or reset the default interval the 5? (`15 / 3`)?

**Comment by michael81877** [comment]
Just FYI i've deployed this branch w/ argo to a remote k8s cluster and verified that the `NifiCluster` and `NifiDataflow` resource are no longer constantly upgrading

image

A screenshot doesn't really prove this, but the resource `resourceVersion` fields are not updating unless I change the CRDs.

**Review by erdrix — APPROVED** [review]

**Pull request merged**